src/index.ts

```diff
@@ -441,6 +441,64 @@ export default class FireStash extends EventEmitter {
     this.eventsTimer = null;
   }
 
+  async onThrottledSnapshot(
+    documentPath: string,
+    callback: (snapshot: FirebaseFirestore.DocumentSnapshot) => void,
+    timeout = 1000,
+  ): Promise<() => void> {
+    // If we've already started the watcher, return.
+    const pending = this.watchers.get(documentPath);
+    if (pending) { return pending; }
+
+    return new Promise((resolve, reject) => {
+      let firstCall: boolean | void = true;
+      let timeoutWindowStart = Date.now();
+      let callsThisTimeout = 0;
+      let numNoChangeTimeouts = 0;
+      let onSnapshotListener: (() => void) | null = null;
+      let interval: NodeJS.Timeout | null = null;
+      let lastUpdateTime = 0;
+
+      const handleSnapshot = async(snapshot: FirebaseFirestore.DocumentSnapshot) => {
+        const now = Date.now();
+        const updateTime = snapshot.updateTime?.toMillis() || 0;
+        const changed = updateTime !== lastUpdateTime;
+        lastUpdateTime = updateTime;
+
+        if (onSnapshotListener) {
+          if (now - timeoutWindowStart >= timeout) {
+            timeoutWindowStart = now;
+            callsThisTimeout = 0;
+          }
+          else callsThisTimeout += 1;
+          if (callsThisTimeout >= 3) {
+            // Cancel the snapshot listener.
+            onSnapshotListener();
+            onSnapshotListener = null;
+            callsThisTimeout = 0;
+            // Switch to polling.
+            interval = setInterval(() => this.db.doc(documentPath).get().then(handleSnapshot, reject), timeout);
+          }
+        }
+        else {
+          if (!changed) ++numNoChangeTimeouts;
+          if (numNoChangeTimeouts >= 3) {
+            // Cancel the interval.
+            interval && clearInterval(interval);
+            numNoChangeTimeouts = 0;
+            onSnapshotListener = this.db.doc(documentPath).onSnapshot(handleSnapshot, reject);
+          }
+        }
+        callback(snapshot);
+        firstCall && (firstCall = resolve());
```

Comment:
resolve() should resolve with a function that when called will clear the interval and cancel the snapshot listener.

Aka, unsubscribe :)

Reply:
Will require one more test to validate unsubscribe functionality.

Reply:
I pulled `unsubscribe()` out to its own function for the cache watcher because it does a little more than just cancel watchers. We're emulating Firebase's onSnapshot here though, so we should mimic its behavior in this API.